internal/faults/faults.go

```diff
@@ -0,0 +1,277 @@
+	serviceLabel       = "fault_service"
+	methodLabel        = "fault_method"
+	protocolLabel      = "fault_protocol"
+	successLabel       = "fault_success"
+	delayInjectedLabel = "fault_injected_delay"
+	abortInjectedLabel = "fault_injected_abort"
+)
+
+var (
+	totalRequests = promauto.With(prometheusbpint.GlobalRegistry).NewCounterVec(prometheus.CounterOpts{
+		Name: "faultbp_fault_requests_total",
+		Help: "Total count of requests seen by the fault injection middleware.",
+	}, []string{
+		clientNameLabel,
+		serviceLabel,
+		methodLabel,
+		protocolLabel,
+		successLabel,
+		delayInjectedLabel,
+		abortInjectedLabel,
+	})
+)
+
+// Headers is an interface to be implemented by the caller to allow
+// protocol-specific header lookup. Using an interface here rather than a
+// function type avoids any potential closure requirements of a function.
+type Headers interface {
+	// Lookup returns the value of a protocol-specific header with the
+	// given key.
+	Lookup(ctx context.Context, key string) (string, error)
+}
+
+// Resume is the function type to continue processing the protocol-specific
+// request without injecting a fault.
+type Resume[T any] func() (T, error)
+
+// Abort is the function type to inject a protocol-specific fault with the
+// given code and message.
+type Abort[T any] func(code int, message string) (T, error)
+
+// The canonical address for a cluster-local address is <service>.<namespace>,
+// without the local cluster suffix or port. The canonical address for a
+// non-cluster-local address is the full original address without the port.
+func getCanonicalAddress(serverAddress string) string {
+	// Cluster-local address.
+	if i := strings.Index(serverAddress, ".svc.cluster.local"); i != -1 {
+		return serverAddress[:i]
+	}
+	// External host:port address.
+	if i := strings.LastIndex(serverAddress, ":"); i != -1 {
+		port := serverAddress[i+1:]
+		// Verify this is actually a port number.
+		if port != "" && port[0] >= '0' && port[0] <= '9' {
+			return serverAddress[:i]
+		}
+	}
+	// Other address, i.e. unix domain socket.
+	return serverAddress
+}
+
+func parsePercentage(percentage string) (int, error) {
+	if percentage == "" {
+		return 100, nil
+	}
+	intPercentage, err := strconv.Atoi(percentage)
+	if err != nil {
+		return 0, fmt.Errorf("provided percentage %q is not a valid integer: %w", percentage, err)
+	}
+	if intPercentage < 0 || intPercentage > 100 {
+		return 0, fmt.Errorf("provided percentage \"%d\" is outside the valid range of [0-100]", intPercentage)
+	}
+	return intPercentage, nil
+}
+
+// Global variable to allow overriding in tests.
+var selected = func(percentage int) bool {
+	// Use a different random integer per feature as per
+	// https://github.com/grpc/proposal/blob/master/A33-Fault-Injection.md#evaluate-possibility-fraction.
+	return rand.IntN(100) < percentage
+}
+
+// Global variable to allow overriding in tests.
+var sleep = func(ctx context.Context, d time.Duration) error {
```

Comment:
I don't know if we have another test that does something similar, but overall I would try to avoid adding global state for testing or adding code to the non-test file to make it testable for that matter.

It can make some of the code or tests harder to understand, but also potentially lead to flaky tests when it comes to parallel execution.

Would it be possible to move the function into the `Injector` type?

Reply:
Done! This was a vestige from some iterations of the code. I agree moving it to the Injector type is much cleaner.